src/ClientHandle.cpp

```diff
@@ -184,7 +188,23 @@ void cClientHandle::Destroy(void)
 		{
 			player->StopEveryoneFromTargetingMe();
 			player->SetIsTicking(false);
-			world->RemovePlayer(player, true);
+
+			if (WasAddedToWorld)
+			{
+				// If ownership was transferred, our own smart pointer should be unset
+				ASSERT(!m_PlayerPtr);
+
+				m_PlayerPtr = world->RemovePlayer(*player, true);
+
+				// And RemovePlayer should have returned a valid smart pointer
+				ASSERT(m_PlayerPtr);
+			}
+			else
+			{
+				// If ownership was not transferred, our own smart pointer should be valid and RemovePlayer's should not
+				ASSERT(m_PlayerPtr);
+				ASSERT(!world->RemovePlayer(*player, true));
```

Comment:
Observe your first reply: "I think the intention was...." - even you're not sure anymore, so the code really smells :)
Yes please, add a debug-only function for this check. It won't be the first nor the last. (*ehm* *ehm* `cCriticalSection::IsLockedByCurrentThread` *ehm* *ehm* )